deepspeed/pt/deepspeed_zero_optimizer.py

```diff
@@ -252,15 +308,27 @@ def __init__(self,
                 self.fp16_groups_flat[i])
             self.parallel_partitioned_fp16_groups.append(data_parallel_partitions)
 
-            # a partition of the fp32 master weights that will be updated by this process
-            self.single_partition_of_fp32_groups.append(
-                self.parallel_partitioned_fp16_groups[i]
-                [partition_id].clone().float().detach())
+            #that this process will update
+            if self.cpu_offload:
+                start = self.single_partition_of_fp32_groups_index
+                self.single_partition_of_fp32_groups_index += self.parallel_partitioned_fp16_groups[
+                    i][partition_id].numel()
+                end = self.single_partition_of_fp32_groups_index
+                #self.single_partition_of_fp32_groups[start:end] = torch.tensor(self.parallel_partitioned_fp16_groups[i][partition_id].detach(),
+                #                                                   dtype=self.single_partition_of_fp32_groups[i].dtype)
+                self.single_partition_of_fp32_groups[

```

Comment:
We do not creating an fp32 copy since we do not want to waste GPU memory space. I test such usage offline with torch.cuda.memory_allocated() and torch.cuda.max_memory_allocated(), and it can convert fp16 to fp32 without extra memory usage